src/snowflake/snowpark/_internal/compiler/common_subdataframe_elimination.py

```diff
@@ -0,0 +1,232 @@
+    subquery generated.
+    For example:
+       df = session.table("test_table")
+       df1 = df1.select("a", "b")
+       df2 = df1.union_all(df1)
+    originally the generated query for df2 is
+        (select "a", "b" from "test_table") union all (select "a", "b" from "test_table")
+    after the optimization, the generated query becomes
+        with temp_cte_xxx as (select "a", "b" from "test_table")
+        (select * from temp_cte_xxx) union all (select * from select * from temp_cte_xxx)
+    """
+
+    # original logical plans to apply the optimization on
+    _logical_plans: List[LogicalPlan]
+    # record the occurrence of each node in a logical plan tree
+    _node_count_map: Dict[TreeNode, int]
+    # record a map between a node and its parents for a logical plan tree
+    _node_parents_map: Dict[TreeNode, Set[TreeNode]]
+    # duplicated node detected for a logical plan
+    _duplicated_nodes: Set[TreeNode]
+    _query_generator: QueryGenerator
+
+    def __init__(
+        self, logical_plans: List[LogicalPlan], query_generator: QueryGenerator
+    ) -> None:
+        self._logical_plans = logical_plans
+        self._query_generator = query_generator
+        self._node_count_map = defaultdict(int)
+        self._node_parents_map = defaultdict(set)
+        self._duplicated_nodes = set()
+
+    def apply(self) -> List[LogicalPlan]:
+        """
+        Applies Common SubDataframe elimination on the set of logical plans one after another.
+
+        Returns
+        -------
+        A set of the new LogicalPlans with common sub dataframe deduplicated with CTE node.
+        """
+        final_logical_plans: List[LogicalPlan] = []
+        for logical_plan in self._logical_plans:
+            # clear the node_count and parents map
+            self._node_count_map = defaultdict(int)
+            self._node_parents_map = defaultdict(set)
+            self._duplicated_nodes.clear()
+
+            # NOTE: the current common sub-dataframe elimination relies on the
+            # fact that all intermediate steps are resolved properly. Here we
+            # do a pass of resolve of the logical plan to make sure we get a valid
+            # resolved plan to start the process.
+            # If the plan is already a resolved plan, this step will be a no-op.
+            logical_plan = self._query_generator.resolve(logical_plan)
+
+            # apply the CTE optimization on the resolved plan
+            self._find_duplicate_subtrees(logical_plan)
+            if len(self._duplicated_nodes) > 0:
+                deduplicated_plan = self._deduplicate_with_cte(logical_plan)
+                final_logical_plans.append(deduplicated_plan)
+            else:
+                final_logical_plans.append(logical_plan)
+
+        # TODO (SNOW-1566363): Add telemetry for CTE
+        return final_logical_plans
+
+    def _find_duplicate_subtrees(self, root: TreeNode) -> None:
```

Comment:
Is it same with the implementaion in cte_utils.py? If yes, maybe import from it first?